File: generated_code/cpp/test_unittest_572.cpp
#include <gtest/gtest.h>
#include <vector>
#include <queue>
#include "solution_proxy.h"

// Helper function to build a binary tree from a level-order vector (using -100000 for null nodes)
TreeNode* buildTree(const std::vector<int>& nodes) {
    if (nodes.empty() || nodes[0] == -100000) return nullptr;
    
    TreeNode* root = new TreeNode(nodes[0]);
    std::queue<TreeNode*> q;
    q.push(root);
    
    size_t i = 1;
    while (!q.empty() && i < nodes.size()) {
        TreeNode* curr = q.front();
        q.pop();
        
        if (i < nodes.size() && nodes[i] != -100000) {
            curr->left = new TreeNode(nodes[i]);
            q.push(curr->left);
        }
        i++;
        
        if (i < nodes.size() && nodes[i] != -100000) {
            curr->right = new TreeNode(nodes[i]);
            q.push(curr->right);
        }
        i++;
    }
    return root;
}

// Helper function to free the allocated binary tree memory
void freeTree(TreeNode* root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    delete root;
}

class IsSubtreeTest : public ::testing::Test {
protected:
    Solution solution;
};

// Test Case 1: Example 1 from problem description (True case)
TEST_F(IsSubtreeTest, Example1_ReturnsTrue) {
    TreeNode* root = buildTree({3, 4, 5, 1, 2});
    TreeNode* subRoot = buildTree({4, 1, 2});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(freeTree); // Note: typo corrected below to subRoot
    freeTree(subRoot);
}

// Test Case 2: Example 2 from problem description (False case due to extra child node)
TEST_F(IsSubtreeTest, Example2_ReturnsFalse) {
    // -100000 represents null nodes in the level-order traversal
    TreeNode* root = buildTree({3, 4, 5, 1, 2, -100000, -100000, -100000, -100000, 0});
    TreeNode* subRoot = buildTree({4, 1, 2});
    
    EXPECT_FALSE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 3: Both trees consist of a single matching node (Minimum constraints)
TEST_F(IsSubtreeTest, SingleNode_Identical_ReturnsTrue) {
    TreeNode* root = buildTree({1});
    TreeNode* subRoot = buildTree({1});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 4: Both trees consist of a single node but have different values
TEST_F(IsSubtreeTest, SingleNode_DifferentValues_ReturnsFalse) {
    TreeNode* root = buildTree({1});
    TreeNode* subRoot = buildTree({2});
    
    EXPECT_FALSE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 5: The subRoot is exactly identical to the entire root tree
TEST_F(IsSubtreeTest, EntireTreeIsSubtree_ReturnsTrue) {
    TreeNode* root = buildTree({10, 5, 15, 3, 7});
    TreeNode* subRoot = buildTree({10, 5, 15, 3, 7});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 6: Deep skewed tree structure where subRoot matches the end of the chain
TEST_F(IsSubtreeTest, DeepSkewedTree_MatchesEnd_ReturnsTrue) {
    TreeNode* root = buildTree({1, 2, -100000, 3, -100000, 4, -100000, 5});
    TreeNode* subRoot = buildTree({4, 5});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 7: The subRoot structure is similar but mirror-inverted
TEST_F(IsSubtreeTest, StructuralMismatch_MirrorImage_ReturnsFalse) {
    TreeNode* root = buildTree({3, 4, 5, 1, -100000});
    TreeNode* subRoot = buildTree({4, -100000, 1});
    
    EXPECT_FALSE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 8: Boundary extreme values for node data (-10^4 and 10^4)
TEST_F(IsSubtreeTest, ExtremeNegativeAndPositiveValues_ReturnsTrue) {
    TreeNode* root = buildTree({-10000, 10000, 0});
    TreeNode* subRoot = buildTree({10000});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

// Test Case 9: Duplicate subtrees present in the main tree
TEST_F(IsSubtreeTest, DuplicateSubtreesInRoot_ReturnsTrue) {
    TreeNode* root = buildTree({1, 2, 2, 4, 5, 4, 5});
    TreeNode* subRoot = buildTree({2, 4, 5});
    
    EXPECT_TRUE(solution.isSubtree(root, subRoot));
    
    freeTree(root);
    freeTree(subRoot);
}

File: generated_code/python/test_unittest_572.py
import os
import importlib.util
import unittest

# Dynamic loading of the solution module as per guidelines
solution_path = os.environ.get("TEST_SOLUTION_FILE")
if not solution_path:
    raise RuntimeError("TEST_SOLUTION_FILE environment variable is not set.")

spec = importlib.util.spec_from_file_location("Solution", solution_path)
sol_module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(sol_module)
Solution = sol_module.Solution
TreeNode = sol_module.TreeNode

class TestIsSubtree(unittest.TestCase):
    
    def _build_tree(self, nodes):
        """Helper method to construct a binary tree from a LeetCode level-order list."""
        if not nodes:
            return None
        
        root = TreeNode(nodes[0])
        queue = [root]
        front = 0
        index = 1
        
        while index < len(nodes):
            node = queue[front]
            front += 1
            
            # Left child
            if nodes[index] is not None:
                node.left = TreeNode(nodes[index])
                queue.append(node.left)
            index += 1
            
            # Right child
            if index < len(nodes):
                if nodes[index] is not None:
                    node.right = TreeNode(nodes[index])
                    queue.append(node.right)
                index += 1
                
        return root

    def test_example_1_positive_match(self):
        # Input: root = [3,4,5,1,2], subRoot = [4,1,2] -> True
        root = self._build_tree([3, 4, 5, 1, 2])
        sub_root = self._build_tree([4, 1, 2])
        
        sol = Solution()
        self.assertTrue(sol.isSubtree(root, sub_root))

    def test_example_2_structural_mismatch(self):
        # Input: root = [3,4,5,1,2,None,None,None,None,0], subRoot = [4,1,2] -> False
        root = self._build_tree([3, 4, 5, 1, 2, None, None, None, None, 0])
        sub_root = self._build_tree([4, 1, 2])
        
        sol = Solution()
        self.assertFalse(sol.isSubtree(root, sub_root))

    def test_identical_trees(self):
        # Core Constraint/Property: A tree is a subtree of itself
        root = self._build_tree([1, 2, 3])
        sub_root = self._build_tree([1, 2, 3])
        
        sol = Solution()
        self.assertTrue(sol.isSubtree(root, sub_root))

    def test_single_node_match(self):
        # Minimum input constraints: Both trees have 1 node and match
        root = self._build_tree([42])
        sub_root = self._build_tree([42])
        
        sol = Solution()
        self.assertTrue(sol.isSubtree(root, sub_root))

    def test_single_node_mismatch(self):
        # Minimum input constraints: Both trees have 1 node but values differ
        root = self._build_tree([42])
        sub_root = self._build_tree([99])
        
        sol = Solution()
        self.assertFalse(sol.isSubtree(root, sub_root))

    def test_subroot_deeper_than_root(self):
        # Structural constraint: subRoot has more nodes than root, cannot be a subtree
        root = self._build_tree([1])
        sub_root = self._build_tree([1, 2, 3])
        
        sol = Solution()
        self.assertFalse(sol.isSubtree(root, sub_root))

    def test_negative_values_match(self):
        # Value boundaries: Constraints specify -10^4 <= val <= 10^4
        root = self._build_tree([-10, -20, -30, -40])
        sub_root = self._build_tree([-20, -40])
        
        sol = Solution()
        self.assertTrue(sol.isSubtree(root, sub_root))

    def test_duplicate_values_in_root_correct_match(self):
        # Ensure the algorithm traverses the entire tree and finds the actual matching subtree 
        # even if an earlier node with the same value fails the full structural match.
        root = self._build_tree([4, 4, 5, 1, 2]) # Top root is 4, left child is also 4
        sub_root = self._build_tree([4, 1, 2])
        
        sol = Solution()
        self.assertTrue(sol.isSubtree(root, sub_root))

if __name__ == "__main__":
    unittest.main()